cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,404 @@
+        fitness_distribution: Optional[Callable[[], float]] = None,
+        fitness_base: float = np.e,
+        num_extant: Optional[int] = None,
+        experiment_time: Optional[float] = None,
+        collapse_unifurcations: bool = True,
+        random_seed: int = None,
+    ):
+        if num_extant is None and experiment_time is None:
+            raise TreeSimulatorError(
+                "Please specify at least one stopping condition"
+            )
+        if mutation_distribution is not None and fitness_distribution is None:
+            raise TreeSimulatorError(
+                "Please specify a fitness strength distribution"
+            )
+        if num_extant is not None and num_extant <= 0:
+            raise TreeSimulatorError(
+                "Please specify number of extant lineages greater than 0"
+            )
+        if num_extant is not None and type(num_extant) is not int:
+            raise TreeSimulatorError(
+                "Please specify an integer number of extant tips"
+            )
+        if experiment_time is not None and experiment_time <= 0:
+            raise TreeSimulatorError(
+                "Please specify an experiment time greater than 0"
+            )
+
+        self.birth_waiting_distribution = birth_waiting_distribution
+        self.initial_birth_scale = initial_birth_scale
+        self.death_waiting_distribution = death_waiting_distribution
+        self.mutation_distribution = mutation_distribution
+        self.fitness_distribution = fitness_distribution
+        self.fitness_base = fitness_base
+        self.num_extant = num_extant
+        self.experiment_time = experiment_time
+        self.collapse_unifurcations = collapse_unifurcations
+        self.random_seed = random_seed
+
+    def simulate_tree(
+        self,
+    ) -> CassiopeiaTree:
+        """Simulates trees from a general birth/death process with fitness.
+
+        A forward-time birth/death process is simulated by tracking a series of
+        lineages and sampling event waiting times for each lineage. Each
+        lineage draws death waiting times from the same distribution, but
+        maintains its own birth scale parameter that determines the shape of
+        its birth waiting time distribution. At each division event, fitness
+        mutation events are sampled, and the birth scale parameter is scaled by
+        their multiplicative coefficients. This updated birth scale passed
+        onto successors.
+
+        Returns:
+            A CassiopeiaTree with the tree topology initialized with the
+            simulated tree
+        """
+
+        # Samples whether birth, death, or the end of the experiment comes next
+        # for a given lineage, and any fitness changes
+        def sample_lineage_event(lineage: Dict[str, Union[int, float]]) -> None:
+            """A helper function that samples an event for a lineage.
+
+            Takes a lineage and determines the next event in that lineage's
```

Comment:
I have a general design question -- what's the motivation for making these nested functions vs class methods? I think using fewer nested functions makes code more readable, but I'm open to other designs if there's a good reason for it.

Reply:
I think I wanted to avoid having to pass around the tree, queue, and name generator to the helper functions. I think that this might be a small price to pay though if we feel it would improve readability. What do you think?

Reply:
Hmm. Perhaps I'm misreading something, but aren't you still passing around these objects? @sprillo what do you think about this?

Reply:
I guess I'm just saying that `sample_lineage_event` would have to take additional parameters. But I think it might be helpful to show exactly what is being altered by `sample_lineage_event`. So I think I'm actually down to make them instance methods and not nested functions

Reply:
Ah I see, I see. Yes, I think it's a bit cleaner but if it's a lot of work with few if any benefits I don't think it's needed.

Reply:
I'll whip it up haha and then let's push!